// pblog message types shared between protocol ends

package pblog.protocol;

// describes an individual stream
message StreamInfo {
    optional string path = 1;
    optional uint64 length = 2;
}

// describes an individual stream set
message StreamSetInfo {
    // path to this stream set
    optional string path = 1;
    
    // describes streams within this set
    repeated StreamInfo stream = 2;
}

// describes an individual bucket
message BucketInfo {
    // path to this bucket
    optional string path = 1;
    
    // describes stream sets within this bucket
    repeated StreamSetInfo stream_set = 2;
}

// describes a stream store
message StoreInfo {
    // describes buckets within the store
    repeated BucketInfo bucket = 1;
}

// describes how to get an individual stream
message GetStreamDescription {
    // path to stream to get
    optional string path = 1;

    // OR

    // components of path to retrieve
    optional string bucket = 2;
    optional string stream_set = 3;
    optional string stream = 4;

    // offsets within stream to fetch, in bytes
    // end_byte_offset is typically discovered by querying the server to see
    // what offsets are available
    optional uint64 start_byte_offset = 5;
    optional uint64 end_byte_offset = 6;
}